rfcs/20200106-prelu/README.md

```diff
@@ -0,0 +1,137 @@
+# Introducing PReLU support in oneDNN
+
+## Introduction
+
+This paper https://arxiv.org/abs/1502.01852 introduced PReLU activation function.
+Request comes from Caffe users: "PReLU is used in some POR topologies, such as
+VNET, SRGAN. It takes most of time if the original implementation, is not good
+optimized. For example VNET's train perf with latest Intel Caffe is 2.4s, PReLU
+(has optimized by openmp) spends 429ms (not including reorder time), 17% of total
+time. If we uses oneDNN ReLU to make projection of the perf, it will come to
+~1.8s." Some other data scientist in discussion mentioned that he encouter PReLU
+issue when he tried to optimize the inference pipeline of a popular face-related
+application - PReLU tooks 47 % of time in his application.
+
+Currently in oneDNN we support ReLU activation function in eltwise operation:
+https://oneapi-src.github.io/oneDNN/dev_guide_eltwise.html. LeakyReLU can be
+achieved by specifying non zero alpha parameter. PReLU is specific variant of
+LeakyReLU, where alpha is learnable parameter. It cames with two flavours:
+- channel-wise PReLU - where number of trained alpha parameters is equal to
+channel number
+- channel-shared PReLU - where there is only one trainable alpha
+
+All popular framworks supports PReLU:
+- Caffe - https://caffe.berkeleyvision.org/tutorial/layers/prelu.html
+- Keras/tensorflow - https://www.tensorflow.org/api_docs/python/tf/keras/layers/PReLU
+- PyTorch - https://pytorch.org/docs/master/generated/torch.nn.PReLU.html
+- MxNet - https://beta.mxnet.io/api/gluon/_autogen/mxnet.gluon.nn.PReLU.html
+
+Example of calculations forward/backward can be found: https://bit.ly/2TXgvzT
+
+## Proposal
+
+1. Option A - create new dedicated primitive PReLU:
+
+``` cpp
+typedef struct {
+    /// The kind of primitive. Used for self-identifying the primitive
+    /// descriptor. Must be #dnnl_prelu.
+    dnnl_primitive_kind_t primitive_kind;
+    /// The kind of propagation. Possible values: #dnnl_forward_training,
+    /// #dnnl_forward_inference, #dnnl_backward
+    dnnl_prop_kind_t prop_kind;
+    /// Source and destination memory descriptor.
+    dnnl_memory_desc_t data_desc;
+    /// Learnable parameter memory descriptor.
+    dnnl_memory_desc_t weights_desc;
+    /// Source and destination gradient memory descriptor.
+    dnnl_memory_desc_t diff_data_desc;
+    /// Learnable parameter gradient memory descriptor.
+    dnnl_memory_desc_t diff_weights_desc;
+
+} dnnl_prelu_desc_t;
+```
+For passing learnable parameters reuse existing defines:
+``` cpp
+#define DNNL_ARG_WEIGHTS number
+#define DNNL_ARG_DIFF_WEIGHTS number
+```
+Examples:
+n = 256, h = 7, w = 7 c=128
+m_data = {{256, 7, 7, 128}, f32, nhwc};
+
+Algorithm:
+- whole-tensor (tensorflow wariant) - weights_data = {{256, 7, 7, 128}, f32, nhwc};
```

Comment:
Seems like an incomplete example. Please prettify. What is `m_data`, why only one bullet for algorithm, etc..

Reply:
Probably should go towards the end of option 1.